solutions/joi-19-cake-circle.mdx

```diff
@@ -0,0 +1,179 @@
+
+Let $\texttt{opt}(l)$ be the optimal right endpoint for some left endpoint. We can observe that $\texttt{opt}$ is a non-decreasing function. This motivates a divide and conquer solution. 
+
+Suppose we know that $\texttt{opt}(i) \in [p, q]$ for $i \in [l, r]$. If $p = q$, then we are done. Otherwise, let $m = \lfloor \frac{l+r}{2}\rfloor$, and compute $\texttt{opt}(m)$. We now know that $\texttt{opt}(i)$ lies in the range $[p, \texttt{opt(m)}]$ for $i \in [l, m)$, and in the range $[\texttt{opt}(m), q]$ for $i \in (m, r]$. We can then recurse. By the [master theorem](https://en.wikipedia.org/wiki/Master_theorem_(analysis_of_algorithms)), this will run in $\tilde{\mathcal{O}}(n)$ time. 
+
+The only step left is to compute the sum of the $m$ largest values in a range. This can be done with a variety of data structures in logarithmic time.
+
+**Time Complexity**: $\mathcal{O}(n\log^2n)$
+
+# Implementation 
+
+<LanguageSection>
+
+<CPPSection>
+  
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+ 
+namespace std {
+ 
+template<class Fun>
+class y_combinator_result {
+	Fun fun_;
+public:
+	template<class T>
+	explicit y_combinator_result(T &&fun): fun_(std::forward<T>(fun)) {}
+ 
+	template<class ...Args>
+	decltype(auto) operator()(Args &&...args) {
+		return fun_(std::ref(*this), std::forward<Args>(args)...);
+	}
+};
+ 
+template<class Fun>
+decltype(auto) y_combinator(Fun &&fun) {
+	return y_combinator_result<std::decay_t<Fun>>(std::forward<Fun>(fun));
+}
+ 
+}  // namespace std
+ 
+class WaveletTree {
+	int n;
+	vector<long long> vmap;
+	vector<vector<int>> tree;
+	vector<vector<long long>> psum;
+ 
+	template<typename I>
+	void build(I begin, I end, int t, int vl, int vr) {
+		if (vl == vr) {
+			int nn = end - begin + 1;
+			psum[t].reserve(nn);
+			psum[t].push_back(0);
+			for (auto it = begin; it != end; it++)
+				psum[t].push_back(psum[t].back() + vmap[*it]);
+			return;
+		}
+		int vm = (vl + vr) / 2, nn = end - begin + 1;
+		tree[t].reserve(nn);
+		psum[t].reserve(nn);
+		tree[t].push_back(0);
+		psum[t].push_back(0);
+		for (auto it = begin; it != end; it++) {
+			tree[t].push_back(tree[t].back() + (*it <= vm));
+			psum[t].push_back(psum[t].back() + vmap[*it]);
+		}
+		auto pivot = stable_partition(begin, end, [vm](int x) { return x <= vm; });
+		build(begin, pivot, t * 2, vl, vm);
+		build(pivot, end, t * 2 + 1, vm + 1, vr);
+	}
+ 
+	long long query(int l, int r, int k, int t, int vl, int vr) {
+		if (vl == vr) 
+			return k * vmap[vl];
+		int vm = (vl + vr) / 2, lv = tree[t][r] - tree[t][l - 1];
+		if (k <= lv){
+			return query(tree[t][l - 1] + 1, tree[t][r], k, t * 2, vl, vm);
+		} else {
+			return psum[t * 2][tree[t][r]] - psum[t * 2][tree[t][l - 1]] +\
+				query(l - tree[t][l - 1], r - tree[t][r], k - lv, t * 2 + 1, vm + 1, vr);
+		}
+	}
+ 
+ 
+public:
+	WaveletTree(vector<long long> vec) { init(vec.begin(), vec.end()); }
+ 
+	template<typename I>
+	void init(I begin, I end) {
```

Comment:
i'd make this private

Reply:
The point of having an init function is so you can reinitialize it, instead of constructing it again and copying it over

Reply:
add a doc comment for that i guess
something like "resets the data structure to it's initial state for future use"